include/pybind11/detail/class.h

```diff
@@ -340,6 +356,32 @@ inline bool deregister_instance(instance *self, void *valptr, const type_info *t
     return ret;
 }
 
+using derived_tp_init_registry_type
+    = std::unordered_map<PyTypeObject *, int (*)(PyObject *, PyObject *, PyObject *)>;
+
+inline derived_tp_init_registry_type *derived_tp_init_registry() {
+    // Intentionally leak the unordered_map:
+    // https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables
+    static auto *singleton = new derived_tp_init_registry_type();
+    return singleton;
+}
+
+// This mechanism was originally developed here:
+// https://github.com/google/clif/commit/7cba87dd8385ab707c98e814ce742eeca877eb9e
+extern "C" inline int tp_init_intercepted(PyObject *self, PyObject *args, PyObject *kw) {
+    assert(PyType_Check(self) == 0);
```

Comment:
How is this condition guaranteed? Looks like we don't ensure it when intercepting tp_init, based on my initial read of the pybind11object_new method.

PS: if it was not for consisitency with pyclif naming, I would recommend calling this safe_tp_init.

Reply:
> How is this condition guaranteed?

It's not. I put the `assert()` here only to be sure my code isn't doing something unexpected. This PR is already TGP tested, which makes me believe it is very unlikely that the `assert()` will ever fire, but if it does, I'll have a concrete situation to fix (as opposed to anticipating/guessing).

> PS: if it was not for consisitency with pyclif naming,

Name changed to `tp_init_with_safety_checks`.

(After this PR is merged I'll go back and change the PyCLIF code accordingly. I also want to backport the weakref-based cleanup.)